### DISCUSSION TITLE
수많은 DTO를 어떻게 관리해야 할까요?

### DISCUSSION THREAD

**Discussion by thornscript** [GU-99/grow-up-pms#35] [개발 토론]
## 📦 1. 요청 별로 DTO를 외부 클래스로 정의하기
```java
@Getter  
@NoArgsConstructor(access = AccessLevel.PRIVATE)  
public class UserCreateRequest {  
    private String username;
    private String password;
    private String nickname;
    private String content;
    // ...
}

@Getter  
@NoArgsConstructor(access = AccessLevel.PRIVATE)  
public class UserProfileUpdateRequest {
    private String nickname;
    private String content;
    private String profileImage;
    // ...
}
```

### 🌟 장점
- **유연성:** 각 DTO가 개별적으로 정의되므로 특정 요구사항에 맞게 쉽게 수정할 수 있습니다.
- **간단한 구조:** 비교적 간단한 구조로 인해 이해하기 쉽고 관리하기도 수월하다는 장점이 있습니다.

### ⚠️ 단점
- **코드 중복:** 공통된 필드와 메서드가 여러 DTO에 반복될 수 있어, 코드 중복이 발생할 수 있습니다.
- **유지보수 부담:** 여러 DTO에서 동일한 변경을 적용해야 할 때, 각각 수정해야 하므로 유지보수가 어려울 수 있습니다. 이는 DTO의 수가 늘어날수록 크게 부각될 수 있습니다.
	- **일관성 문제:** 각 DTO에서 유효성 검사를 개별적으로 수행하면, 일관성을 유지하기 어렵고 DB에서 엄밀한 제약 조건이 걸려 있지 않은 경우 무결성이 깨질 수 있습니다.
	- **필드 이름 불일치:** 각 DTO에서 필드 이름을 다르게 정의하면, 코드의 일관성이 떨어지며, 유지보수 시 혼란을 초래할 수 있습니다. 또한, 클라이언트와 서버 간의 커뮤니케이션에서 혼란을 초래할 수도 있습니다.

## 📂 2. 내부 클래스를 통해 DTO를 관리하기
```java
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class UserRequest {

    @Getter
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Create {
        private String username;
        private String password;
        private String nickname;
        private String content;
        // ...
    }

    @Getter
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class ProfileUpdate {
        private String nickname;
        private String content;
        private String profileImage;
        // ...
    }
}
```

장단점은 요청 별로 DTO를 외부 클래스로 정의하는 방법과 거의 동일하지만 아래와 같은 부분에서 다소 차이가 있습니다.

### 🌟 장점
- **가독성과 응집도:** 내부 클래스를 사용하면 관련된 클래스를 하나의 파일에 유지할 수 있으므로 응집도가 높아지고 코드가 더 읽기 쉬워집니다.

### ⚠️ 단점
- **지원 문제:** 간혹 라이브러리나 프레임워크에서 내부 클래스를 찾지 못하는 등의 오류가 발생할 수 있습니다(예: Swagger, JPA 등).
- **파일 크기 증가:** 하나의 클래스 파일이 점점 비대해질 수 있어, 관리하기 어렵고 가독성이 떨어질 수 있습니다. 특히, 내부 클래스가 많아지거나 관리하는 필드가 많을수록 파일이 복잡해질 수 있습니다.

## 🏗️ 3. 추상 클래스를 통해 DTO를 공통화하기
```java
@Getter
@SuperBuilder
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public abstract class UserBaseRequest {
    protected String nickname;
    protected String content;
}

@Getter
@SuperBuilder
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class UserCreateRequest extends UserBaseRequest {
    private String username;
    private String password;
}
```

### 🌟 장점
- **코드 중복 제거:** 공통된 필드와 메서드를 추상 클래스에 정의하여 여러 DTO에서 중복된 코드를 줄일 수 있습니다.
- **일관성 유지:** 공통된 부분을 추상 클래스에서 정의함으로써, 각 DTO 간에 일관성을 유지하기 쉽습니다. 이는 특히 많은 DTO가 있을 때 유용합니다.
- **유지보수 용이:** 공통된 부분을 한 곳에서 관리하므로, 변경 사항을 추상 클래스에만 적용하면 되기 때문에 유지보수가 용이합니다.
### ⚠️ 단점
- **유연성 감소:** 모든 DTO가 공통된 구조를 강제받기 때문에, 개별 DTO의 유연성이 떨어질 수 있습니다.
- **의존성 문제:** 추상 클래스의 변경이 해당 클래스를 상속받는 모든 DTO에 영향을 미칠 수 있으므로, 추상 클래스를 변경할 때 주의가 필요합니다.
- **필요하지 않은 필드 포함 가능성**: 추상 클래스가 커질수록 모든 하위 클래스에 필요하지 않은 필드가 포함될 가능성이 높아지고, 이는 불필요한 복잡성과 오버헤드를 초래할 수 있습니다. 그렇다고 추상 클래스가 공통화하는 부분이 너무 작으면 추상 클래스의 장점이 상쇄될 수 있습니다.

## 🔚 결론
DTO를 정의하는 방법에는 여러 가지가 존재하고, 각각의 방법은 저마다의 장단점을 지니고 있습니다. DTO를 설계할 때의 정답은 존재하지 않고, 각 프로젝트의 특성이나 규모, 요구사항에 맞춰서 가장 적합한 방법을 택하는 것이 중요하다는 생각이 들었습니다.

개인적인 의견으로는 대부분의 프로젝트에서는 1번과 같이 필요할 때마다 그 요청에 맞는 DTO를 생성하는 방법을 선호하는 것 같은데, 1번의 방법으로 시작을 했다가 필요에 따라서 2번이나 3번의 방법을 유동적으로 고려해보면 좋겠다는 생각이 들었습니다 🐱 

## 📚 참고 링크
- [DTO를 inner class로 관리하기](https://unluckyjung.github.io/dev/2022/02/20/Dto-InnerClass/)
- [Inner Class로 DTO 관리](https://lackofwillpower.tistory.com/166)
- [[스택 트레이스] DTO Inner Class 사용 시 주의사항](https://hinweis.tistory.com/89)

**Comment by longBright** [comment]
저의 경우, 지금까지는 1번의 방법을 사용해서 요청별, 계층별로 DTO를 관리하는 방식을 사용해왔습니다. DTO를 관리하는 방법이 이렇게 다양하다는 것이 흥미롭네요ㅎ 혹시 계층별 DTO를 사용해본 경험이 없으시다면, 저희 프로젝트에서도 계층별 DTO를 적용해보는건 어떨까요? 저는 개인적으로 각 계층별로 분리하는 방법을 선호합니다. 계층별 분리도 확실해지고, '이론적'으로 깔끔한 코드가 나오기는 하거든요. 아래는 예시입니다!
![image](https://github.com/user-attachments/assets/3e312726-6e53-4ee6-a341-4d5043d38299)

**Reply by thornscript** [reply]
@longBright 기존에는 요청과 관련된 DTO(~Request) 같은 경우에는 프레젠테이션 레이어에서 도메인 레이어로 흐르고, 응답과 관련된 DTO(~Response)는 가능하면 도메인 레이어에서 프레젠테이션 레이어로 흐르도록 구성하고 있었습니다! 혹시 계층별로 DTO를 관리하는 예시 코드나 참고 링크가 있을까요?!

**Reply by longBright** [reply]
저는 해당 관리법을 알게된게 같이 SSAFY 교육 듣는 동기를 통해서 알게된 내용이라, 참고했던 자료는 안남아있지만 방금 찾은거 몇개 공유해드릴게요. 아래 링크들을 참조하시면, 사용 이유나 어떤 식으로 사용하는지에 대한 개요를 잡는데에는 도움이 되실 것 같아요. 더 필요하시면 제가 진행했던 프로젝트 패키지 구조 보여드리면서 설명 드릴게요!

- [Controller와 Service 레이어 간 요청 & 응답 Dto 사용](https://ksh-coding.tistory.com/102)
- [인프런 Q&A](https://www.inflearn.com/questions/944019/controller-service%EC%9A%A9-dto%EB%A5%BC-%EB%B6%84%EB%A6%AC%EC%8B%9C%ED%82%A4%EB%8A%94-%EA%B2%83%EC%97%90-%EB%8C%80%ED%95%9C-%EC%A7%88%EB%AC%B8)
- [Presentation - Business DTO를 분리](https://m-falcon.tistory.com/711)

**Reply by thornscript** [reply]
@longBright 이런 식으로 기존 DTO의 구조를 개선해보는 것도 좋은 경험인 것 같습니다! :+1:  이런건 가급적 빠른 시일 내에 통일을 하고 컨벤션에도 반영하고 작업하면 좋을 것 같다는 생각이 들었는데, 서비스에 특화된 DTO를 보통 어떻게 작명하시나요? `~Command`라고는 되어있는데 영환님이 따로 사용하시는 괜찮은 작명법이 있을까요? :smile:

**Reply by longBright** [reply]
저는 클래스명을 길게 하는걸 별로 안 좋아하기도 하고 계층별로 dto 패키지를 따로 뒀어서, 그냥 `~Dto` 라고 명명했었거든요. 다른 작명법도 고민해봐도 좋을 것 같습니다!

**Reply by thornscript** [reply]
@longBright 서비스 레이어에 특화된 DTO는 `~Request`, `~Response` 아니면 그와 비슷하게 긴 `~Command` 보다는 `~Dto`로 정해보는 것도 좋은 것 같습니다! 향후에 이 부분에 대해서 재현님께 의견을 여쭤보고 이름을 정한 후에 문서에 반영하는 것도 좋을 것 같아요! 좋은 정보 공유해주셔서 감사합니다 :smile:

**Reply by longBright** [reply]
네네 ㅎ 컨트롤러에 Request, Response를 두고 서비스에 Dto를 두는 방식으로 관리했었어요. 논의해보면 될 것 같습니다! 😄

**Reply by KimJayhyun** [reply]
글로 정리하신 내용과 정보 공유해주신 걸로 이해를 잘 했습니다. 

대화 나누신 방향대로,
컨트롤러에 Request, Response를 두고, 서비스에 Dto를 두는 방식으로 개발하겠습니다!
